### ISSUE TITLE
Url::as_str() cannot be parsed back into Url

### ISSUE THREAD

**Issue by e00E** [servo/rust-url#661]
While fuzzing with  https://github.com/servo/rust-url/pull/660 I discovered this failing input

```rust
#[test]
fn fuzzer_error() {
    let parsed = Url::parse("h:/./\r/:").unwrap();
    dbg!(&parsed);
    /*
    Url {
        scheme: "h",
        host: None,
        port: None,
        path: "//:",
        query: None,
        fragment: None,
        username: "",
        password: None,
    }
    */
    let as_str = parsed.as_str();
    dbg!(as_str);
    // "h://:"

    // Err(ParseError::EmptyHost)
    let parsed_again = Url::parse(as_str).unwrap();
    assert_eq!(parsed, parsed_again);
}
```

**Comment by qsantos** [comment]
It looks like this was resolved by #817.

**Comment by valenting** [comment]
Thanks, @qsantos !
